runtime/gc_glue_java/ArrayletObjectModel.cpp

```diff
@@ -139,19 +149,67 @@ GC_ArrayletObjectModel::fixupInternalLeafPointersAfterCopy(J9IndexableObject *de
 	if (hasArrayletLeafPointers(destinationPtr)) {
 		GC_ArrayletLeafIterator leafIterator((J9JavaVM*)_omrVM->_language_vm, destinationPtr);
 		GC_SlotObject *leafSlotObject = NULL;
-		UDATA sourceStartAddress = (UDATA) sourcePtr;
-		UDATA sourceEndAddress = sourceStartAddress + getSizeInBytesWithHeader(destinationPtr);
+		uintptr_t sourceStartAddress = (uintptr_t) sourcePtr;
+		uintptr_t sourceEndAddress = sourceStartAddress + getSizeInBytesWithHeader(destinationPtr);
 
 		while (NULL != (leafSlotObject = leafIterator.nextLeafPointer())) {
-			UDATA leafAddress = (UDATA)leafSlotObject->readReferenceFromSlot();
+			uintptr_t leafAddress = (uintptr_t)leafSlotObject->readReferenceFromSlot();
 
 			if ((sourceStartAddress < leafAddress) && (leafAddress < sourceEndAddress)) {
-				leafSlotObject->writeReferenceToSlot((J9Object*)((UDATA)destinationPtr + (leafAddress - sourceStartAddress)));
+				leafSlotObject->writeReferenceToSlot((J9Object*)((uintptr_t)destinationPtr + (leafAddress - sourceStartAddress)));
 			}
 		}
 	}
 }
 
+bool
+GC_ArrayletObjectModel::isArrayletDataAdjacentToHeader(J9IndexableObject *arrayPtr)
+{
+	uintptr_t dataSizeInBytes = getDataSizeInBytes(arrayPtr);
+	return isArrayletDataAdjacentToHeader(dataSizeInBytes);
+}
+
+bool
+GC_ArrayletObjectModel::isArrayletDataAdjacentToHeader(uintptr_t dataSizeInBytes)
+{
+	MM_GCExtensionsBase *extensions = MM_GCExtensionsBase::getExtensions(_omrVM);
+	uintptr_t minimumSpineSizeAfterGrowing = extensions->getObjectAlignmentInBytes();
+	return ((UDATA_MAX == _largestDesirableArraySpineSize) || (dataSizeInBytes <= (_largestDesirableArraySpineSize - minimumSpineSizeAfterGrowing - contiguousIndexableHeaderSize())));
+}
+
+bool
+GC_ArrayletObjectModel::isAddressWithinHeap(MM_GCExtensionsBase *extensions, void *address)
+{
+	uintptr_t heapBase = (uintptr_t)extensions->heap->getHeapBase();
+	uintptr_t heapTop = (uintptr_t)extensions->heap->getHeapTop();
+	return ((uintptr_t)address >= heapBase) && ((uintptr_t)address < heapTop);
+}
+
+bool
+GC_ArrayletObjectModel::isIndexableObjectDoubleMapped(MM_GCExtensionsBase *extensions, J9IndexableObject *arrayPtr)
+{
+#if defined(J9VM_ENV_DATA64)
+	void *dataAddr = getDataAddrForIndexableObject(arrayPtr);
+	bool isObjectWithinHeap = isAddressWithinHeap(extensions, dataAddr);
+	return ((getDataSizeInBytes(arrayPtr) >= _omrVM->_arrayletLeafSize) && (!isObjectWithinHeap));
+
+#else /* defined(J9VM_ENV_DATA64) */
+	return false;
+#endif /* defined(J9VM_ENV_DATA64) */
+}
+
+bool
+GC_ArrayletObjectModel::shouldFixupDataAddr(MM_GCExtensionsBase *extensions, J9IndexableObject *arrayPtr)
+{
+#if defined(J9VM_ENV_DATA64)
+	void *dataAddr = getDataAddrForIndexableObject(arrayPtr);
```

Comment:
I think it would be more efficient to do a check  similar to what setDataAddrForContiguous does: check is dataAddr points to the end of the header. You may introduce a helper like isDataAddrSetForContiguous (or suggest some other name)

Reply:
... but  I guess, you use this against already moved object, so that would not work. Let me think a bit more...

Reply:
I see, yes this should work if I modify my uses of `shouldFixupDataAddr` to use the old location of the object as opposed to the new one.

Reply:
I hope we could base the check on source object.

In these situations, specifically with evacuating GCs that 'damages' source object, we generally rely on getPreserved***(ForwardedHeader) APIs that gets the information from either ForwardedHeader struct (that took a snapshot of original source object header data before forwarding pointer was installed) or from source object itself if the data was intact. An example is getPreservedIndexableSize. 

I believe that dataAddr will not ever be overwritten by forwarded pointer. For example for CR, FP (which is still uncompressed) will destroy first 64bits of source object, and that is exactly size of class and size fields.

Sliding collector may have its own challenges (may need to fetch info pre move to consume post move).

We can discuss specific API details directly...